cmd/ctl/pkg/convert/convert.go

```diff
@@ -0,0 +1,254 @@
+	err = o.FilenameOptions.RequireFilenameOrKustomize()
+	if err != nil {
+		return err
+	}
+
+	// build the printer
+	o.Printer, err = o.PrintFlags.ToPrinter()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Run executes version command
+func (o *Options) Run() error {
+	builder := new(resource.Builder)
+
+	r := builder.Unstructured().LocalParam(true).ContinueOnError().
+		FilenameParam(false, &o.FilenameOptions).Flatten().Do()
+
+	err := r.Err()
+	if err != nil {
+		return fmt.Errorf("error here: %s", err)
+	}
+
+	singleItemImplied := false
+	infos, err := r.IntoSingleItemImplied(&singleItemImplied).Infos()
+	if err != nil {
+		return fmt.Errorf("error here instead: %s", err)
+	}
+
+	if len(infos) == 0 {
+		return fmt.Errorf("no objects passed to convert")
+	}
+
+	var specifiedOutputVersion schema.GroupVersion
+	if len(o.OutputVersion) > 0 {
+		specifiedOutputVersion, err = schema.ParseGroupVersion(o.OutputVersion)
+		if err != nil {
+			return err
+		}
+	}
+
+	factory := serializer.NewCodecFactory(api.Scheme)
+	serializer := apijson.NewSerializerWithOptions(apijson.DefaultMetaFactory, api.Scheme, api.Scheme, apijson.SerializerOptions{})
+	encoder := factory.WithoutConversion().EncoderForVersion(serializer, nil)
+	objects, err := asVersionedObject(infos, !singleItemImplied, specifiedOutputVersion, encoder)
+	if err != nil {
+		return err
+	}
+
+	return o.Printer.PrintObj(objects, o.Out)
+}
+
+// asVersionedObject converts a list of infos into a single object - either a List containing
+// the objects as children, or if only a single Object is present, as that object. The provided
+// version will be preferred as the conversion target, but the Object's mapping version will be
+// used if that version is not present.
+func asVersionedObject(infos []*resource.Info, forceList bool, specifiedOutputVersion schema.GroupVersion, encoder runtime.Encoder) (runtime.Object, error) {
+	objects, err := asVersionedObjects(infos, specifiedOutputVersion, encoder)
+	if err != nil {
+		return nil, err
+	}
+
+	var object runtime.Object
+	if len(objects) == 1 && !forceList {
+		object = objects[0]
+	} else {
+		object = &metainternalversion.List{Items: objects}
+
+		targetVersions := []schema.GroupVersion{}
+		if !specifiedOutputVersion.Empty() {
+			targetVersions = append(targetVersions, specifiedOutputVersion)
+		}
+		targetVersions = append(targetVersions, schema.GroupVersion{Group: "", Version: "v1"})
+
+		converted, err := tryConvert(object, targetVersions...)
+		if err != nil {
+			return nil, err
+		}
+
+		object = converted
+	}
+
+	actualVersion := object.GetObjectKind().GroupVersionKind()
+
+	if actualVersion.Version != specifiedOutputVersion.Version {
+		defaultVersionInfo := ""
+		if len(actualVersion.Version) > 0 {
+			defaultVersionInfo = fmt.Sprintf("Defaulting to %q", actualVersion.Version)
+		}
+		klog.V(1).Infof("info: the output version specified is invalid. %s\n", defaultVersionInfo)
```

Comment:
I'm a bit confused by this section - what's it checking/doing? How come it detects that the output version is invalid? 🤔

Reply:
Changing this to fmt.printf you will get the following output without defining any target version

```$ ./ctl convert -f  ../../cert.yaml
info: the output version specified is invalid. Defaulting to "v1"
apiVersion: v1
items:
- apiVersion: v1
  kind: Namespace
  metadata:
    creationTimestamp: null
    name: sandbox
  spec: {}
  status: {}
- apiVersion: cert-manager.io/v1alpha2
  kind: Issuer
  metadata:
    creationTimestamp: null
    name: selfsigned-issuer
    namespace: sandbox
  spec:
    selfSigned: {}
  status: {}
- apiVersion: cert-manager.io/v1alpha2
  kind: Certificate
  metadata:
    creationTimestamp: null
    name: ca-issuer
    namespace: sandbox
  spec:
    commonName: my-csi-app
    isCA: true
    issuerRef:
      group: cert-manager.io
      kind: Issuer
      name: selfsigned-issuer
    secretName: ca-key-pair
  status: {}
- apiVersion: cert-manager.io/v1alpha2
  kind: Issuer
  metadata:
    creationTimestamp: null
    name: ca-issuer
    namespace: sandbox
  spec:
    ca:
      secretName: ca-key-pair
  status: {}
- apiVersion: cert-manager.io/v1alpha2
  kind: Certificate
  metadata:
    creationTimestamp: null
    name: ca-issuer-2
    namespace: sandbox
  spec:
    commonName: my-csi-app
    isCA: true
    issuerRef:
      group: cert-manager.io
      kind: Issuer
      name: ca-issuer
    secretName: ca-key-pair
  status: {}
kind: List
metadata: {}
```

Reply:
Hm, is the input in this example actually valid? It includes a `v1` Namespace object too - do we just ignore these? I don't recall what the decision/resolution was on this

Reply:
That's what kubctl is doing. If we haven't specified a target version and we don't recognize the version, we just use what's given